### PR TITLE
REST API: Web Authorization

### LINKED ISSUE #8599: [REST API] Offer WebView authorization as a fallback if the native site credentials login fails.
When the native site credentials login fails, we need to offer the WebView authorization as a fallback solution, and the flow will be:

1. Get the Authorization URL from the fetched site.
2. Open the URL in a WebView, while offering the following:
    i. A redirect URL.
    ii. The Application Name.
    iii. A UUID as the application ID.
3. If the authorization succeeds, then extract the username and password from the redirect URL, then save them to the `ApplicationPasswordsStore`.
4. Then continue the login checks.

### PR DESCRIPTION
**Pull request by hichamboushaba**
Closes: #8599 

### Description
This PR implements the Application Password's web authorization flow.

I suggest reviewing the PR by commits, it'll help understanding the reasoning for each decision.

Previously, REST API login only supports native login, as outlined in the diagram below (click for bigger version):

REST API login flow i1 complete

Behind the scene, this process requires authenticating via `wp-login.php` to get the nonce required to generate the application password. With this process, we're seeing a decent amount of login failures caused by inaccessible `wp-login.php` (due to security plugins or server settings, or other causes).

This PR is meant to help with those login failures by adding an alternative login method using webview, if a merchant gets login failure using the native login.

--- 
#### This PR adds the following:
1. A dialog message once the native REST API login fails,
2. An option in the dialog to login via wp-admin, using the authentication URL provided by WP's REST API,
3. A webview to allow such login action,
4. Once webview authentication is completed, handle the automatic redirect back into the app and log the user in.
5. Error handlings.

This expands the diagram into something like below, with webview login added to the flow starting from the login error on top right:

REST API login flow complete simplified

---

### Not in this PR:
1. Unit tests related to this feature. To keep the PR small-ish, these are added separately in https://github.com/woocommerce/woocommerce-android/pull/8651
2. Tracks. To keep the PR small-ish, these are added separately.
3. Handling for this edge case issue #8644. It will be tackled later on.

--- 

#### Success path: ####

1. Have a test .org WooCommerce site with no Jetpack,
2. Replicate the native REST API login error by doing the following: 
    1. Install **"WP Hide & Security Enhancer"** plugin and activate it,
    2. Go to **"WP Hide"** -> **"Hide > Admin"** in the menu,
    3. On the **"New wp-login.php"** section, enter a custom login url (e.g: "entering.php"),
    4. On the **"Block default wp-login.php"** section, select "Show", then select "Yes"
    5. Click "Save"
3. Start the app,
4. Enter the test site's URL, then enter credentials, and wait until the login failure dialog happens,
5. Select "Try again with wp-admin page",
6. Wait until webview loads, and proceed with the login process,
7. When wp-admin asks you to approve the application passwords connection, please approve it,
8. Wait and confirm that the app logs you in properly.

#### Failure path 1 : Application passwords authentication rejected ####
1. Follow success path steps 1-6,
2. When wp-admin asks you to approve the application passwords connection, please choose "No, I do not approve of this connection",
3. Confirm that the Log in screen is now loaded with a Snackbar explaining the error.

#### Failure path 2: Application passwords disabled ####
This for the case where both `wp-login.php` is inaccessible, and application password is disabled.
1. Follow success path steps 1-2, 
3. In wp-admin, install and activate "Disable Application Passwords" plugin,
4. Follow success path 3-5,
6. Confirm that an error page is shown saying that application password is not available.

### Video

Check video below for a successful login example:

https://user-images.githubusercontent.com/266376/227840023-e303794d-3381-412d-810d-53313a427762.mov

---

Check the screenshots below for the failure cases:

| Application passwords not approved  | Application passwords disabled |
|-|-|
|  application password not approved | Screenshot_20230327_112531 |

- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by hafizrahman** [comment]
Setting as "Do not merge" for now until we decide whether to put this into beta or the next release.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/sitecredentials/LoginSiteCredentialsViewModel.kt [hunk 1]
```diff
@@ -66,22 +77,18 @@ class LoginSiteCredentialsViewModel @Inject constructor(
...
-    ) { siteAddress, username, password, isLoading, errorDialog ->
-        LoginSiteCredentialsViewState(
-            siteUrl = siteAddress,
-            username = username,
-            password = password,
-            isLoading = isLoading,
-            errorDialogMessage = errorDialog
-        )
+    @OptIn(ExperimentalCoroutinesApi::class)
+    val viewState = state.flatMapLatest {
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/sitecredentials/LoginSiteCredentialsViewModel.kt:84]: Np, but the field `state` I'd maybe rename it to `initialState` or something similar to clearly be able to differentiate from the actual `viewState`. 

Very nice usage of Flows API in this view model.

**Review by JorgeMucientes — APPROVED** [review]

**Comment by JorgeMucientes** [comment]
Outstanding job here @hichamboushaba and @hafizrahman 🥇. This wasn't an easy task to deal with! The code looks great and works like a charm for the different testing scenarios. 

I just found a minor corner case where the UI state for `LoginSiteCredentialsViewModel` is not restored properly. Basically, when clicking on arrow back from "Application passwords disabled" error screen. See the screen recording: 


https://user-images.githubusercontent.com/2663464/228500586-be90e003-11ec-4311-9d5b-88364da9d6fe.mp4

Since it's a minor thing, I wouldn't block the PR because of that. So, approved. Let's merge it after the tracking PR is merged into this branch.

**Comment by hafizrahman** [comment]
@JorgeMucientes Thanks for the review and bug report! Great minds think alike, I've also seen that bug and written about it in #8644. 

I'll fix it in the next update since I think it's quite an edge case. It looks awkward but hopefully not many users disable both wp-login.php access and application passwords at the same time. In the meanwhile they also can still tap the "X" button to close it. Again, very awkward, but not end of the world bug.

**Comment by JorgeMucientes** [comment]
> I'll fix it in the next update since I think it's quite an edge case.

Agreed! Sounds good. 

Btw while re-testing the feature I found a similar corner case.  For "Failure path 2: Application passwords disabled" from the PR description: 
- Try to log in with app paswords disabled
- See error page is shown saying that the application password is not available.
- Go to `wc-admin` disable the plugin "Disable Application Passwords"
- Click "Retry" button in the app from the application passwords disabled error screen 
- Confirm the blank screen


https://user-images.githubusercontent.com/2663464/228665131-0015a032-acfe-48b3-9e9e-cda9c79cd851.mp4

This is also a non blocking corner case but maybe you want to keep track of it.

**Comment by hafizrahman** [comment]
Just to note that between these last few commits, I've seen check failures related to e2e test, especially in order creation flow. I did some testing and couldn't replicate it. There is nothing in this PR that relates to that issue, and in last commit https://github.com/woocommerce/woocommerce-android/pull/8618/commits/7a71d9ce01979230dd1949352d9c87e69c0b3ee8 the failure disappeared on its own without any specific fixing from this PR. So in case this happens again in the next commits, I'd say it should be OK to ignore.

**Comment by hafizrahman** [comment]
> Btw while re-testing the feature I found a similar corner case. For "Failure path 2: Application passwords disabled" from the PR description:

@JorgeMucientes nice find, thanks for the report! I've added a fix in https://github.com/woocommerce/woocommerce-android/pull/8618/commits/5df0192433cd660aa220855226655da1b35a21c0

**Pull request merged**